### DISCUSSION TITLE
How to achieve multiline text with char wrapping?

### DISCUSSION THREAD

**Discussion by yanshay** [slint-ui/slint#6284] [Q&A] [answered]
When using Text and separating strings with '\n' every string shows on a new line.
When setting `wrap` to `char-wrap` all strings show one after the other with no separation to lines.

Is there a way to both have wrapping of text and separate lines (in a single text)?

**Comment by Enyium** [comment]
The [docs](https://releases.slint.dev/1.8.0/docs/slint/src/language/builtins/enums#textwrap) for `char-wrap` say:

> Currently only supported by the Qt and Software renderers.

Is your app compliant with that? Default is FemtoVG.

**Reply by ogoffart** [reply]
true, this reminds me we don't have an issue to track the implementation, so i have filled https://github.com/slint-ui/slint/issues/6292

**Reply by yanshay** [reply]
I'm using the software-rendered on MCU. 
It does do the wrapping, only the new-lines are not respected and multiline text becomes a long wrapped line and can't be read, so it is not usable for multiline text.

**Comment by ogoffart** [comment]
Would you have an example of Slint code that shows the wrong behavior, and what behavior you're expecting?

**Reply by yanshay** [reply]
The following code w/o the `wrap-char`:
```
                Rectangle {
                  Text {
                    width: root.width;
                    height: root.height;
                    color: white;
                    text: "This is a long line, longer than the display can handle, so I'd want it to wrap\nAnd this is the line after it which I'd like to see starting on a new line and then wrapping on char boundary";
                    font-size:20px;
                    vertical-alignment: top;
                    horizontal-alignment: left;
                  }
                }
```
**Image1: On the device it shows as:**

![IMG_8209](https://github.com/user-attachments/assets/aee60b9c-aa6c-4899-9d9f-4625ac293e8e)


**Image2: In Live Preview it appears as**
![image](https://github.com/user-attachments/assets/f006a6c7-1d2f-4257-a1f9-cd5a5632f029)

**Image3: When I add `wrap: char-wrap`** 
![IMG_8210](https://github.com/user-attachments/assets/e2478260-fa18-40fd-8d50-307dd711ec62)

In live preview `wrap-char` has no effect.

What I'd expect is:
1. Image-1 is fine w/o when not requested to wrap.
2. I'd expect the char-wrap to wrap the lines at the char level (not word unless very large words), which the software-renderer does ok as you can see on Image-3, BUT I'd expect the new lines to starts on new lines like it is in image-2 from the live-preview (which wraps even when not instructed to) and not just continue right after the previous line as it currently is as can be seen on Image-3
1. I think it's important to have consistency across renderers.

**Reply by ogoffart** [reply]
Indeed, thanks for the picture. I understand now. There is indeed a bug in the software renderer.
You may get better result with `wrap: word-wrap`

**Reply by ogoffart** [reply, marked as answer]
Reported as bug in https://github.com/slint-ui/slint/issues/6346

**Reply by yanshay** [reply]
> Reported as bug in #6346

And I see also fixed. 
Awesome, thanks!
